src/interpreter.rs

```diff
@@ -1233,79 +1233,93 @@ impl SignatureChecker for CallbackTransactionSignatureChecker<'_> {
     }
 }
 
-/// Additional validation for spend-to-script-hash transactions:
-fn eval_p2sh<F>(
-    data_stack: Stack<Vec<u8>>,
+fn eval_sig<F>(
     script_sig: &script::Code,
+    flags: VerificationFlags,
     payload: &mut F::Payload,
     stepper: &F,
 ) -> Result<Stack<Vec<u8>>, script::Error>
 where
     F: StepFn,
 {
-    // script_sig must be literals-only or validation fails
-    if script_sig.is_push_only() {
-        // stack cannot be empty here, because if it was the P2SH HASH <> EQUAL scriptPubKey would
-        // be evaluated with an empty stack and the `eval_script` in the caller would return false.
-        assert!(!data_stack.is_empty());
-        data_stack
-            .split_last()
-            .map_err(|e| script::Error::Interpreter(None, e))
-            .and_then(|(pub_key_2, remaining_stack)| {
-                eval_script(remaining_stack, &script::Code(pub_key_2), payload, stepper)
-            })
-            .and_then(|p2sh_stack| {
-                if p2sh_stack.last().is_ok_and(cast_to_bool) {
-                    Ok(p2sh_stack)
-                } else {
-                    Err(script::Error::EvalFalse)
-                }
-            })
-    } else {
+    if flags.contains(VerificationFlags::SigPushOnly) && !script_sig.is_push_only() {
         Err(script::Error::SigPushOnly)
+    } else {
+        eval_script(Stack::new(), script_sig, payload, stepper)
     }
 }
 
+/// Additional validation for spend-to-script-hash transactions:
+fn eval_p2sh<F>(
+    data_stack: Stack<Vec<u8>>,
+    payload: &mut F::Payload,
+    stepper: &F,
+) -> Result<Stack<Vec<u8>>, script::Error>
+where
+    F: StepFn,
+{
+    // stack cannot be empty here, because if it was the P2SH HASH <> EQUAL scriptPubKey would
```

Comment:
Why is it that we no longer need the push-only check here? The commit message on this commit is not very informative with respect to this change. Is it that this is only invoked from locations where the check has already been performed?

Reply:
The check is still done, but it remains in `verify_script`. This is because it applies to the `Sig` component, not the `Redeem` component.